### PR TITLE
IGNITE-19999 Sql. Add planner test to verify numeric type coercion in binary comparison

### PR DESCRIPTION
**Pull request by korlov42**

### CODE REVIEW

File: modules/sql-engine/src/test/java/org/apache/ignite/internal/sql/engine/planner/AbstractPlannerTest.java [hunk 1]
```diff
@@ -570,6 +570,10 @@ protected <T extends RelNode> void assertPlan(
                 return;
             }
         } catch (Throwable th) {
+            if (lastErrorMsg == null) {
```
Comment (ygerzhedovich) [inline modules/sql-engine/src/test/java/org/apache/ignite/internal/sql/engine/planner/AbstractPlannerTest.java:573]: is it last or first?

File: modules/sql-engine/src/test/java/org/apache/ignite/internal/sql/engine/planner/datatypes/NumericComparisonTypeCoercionTest.java [hunk 2]
```diff
@@ -0,0 +1,965 @@
...
+                        .secondOpMatches(castTo(NativeTypes.DOUBLE)),
+
+                forTypePair(NumericPair.DECIMAL_5_0_DOUBLE)
+                        .firstOpMatches(castTo(NativeTypes.DOUBLE))
+                        .secondOpMatches(ofTypeWithoutCast(NativeTypes.DOUBLE)),
+
+
+                forTypePair(NumericPair.DECIMAL_6_1_DECIMAL_6_1)
+                        .firstOpMatches(ofTypeWithoutCast(Types.DECIMAL_6_1))
+                        .secondOpMatches(ofTypeWithoutCast(Types.DECIMAL_6_1)),
```
Comment (AMashenkov) [inline modules/sql-engine/src/test/java/org/apache/ignite/internal/sql/engine/planner/datatypes/NumericComparisonTypeCoercionTest.java:724]: This looks similar to NumericPair.DECIMAL_8_3_DECIMAL_8_3 and NumericPair.DECIMAL_5_3_DECIMAL_5_3.

**Review by AMashenkov — APPROVED** [review]

File: modules/sql-engine/src/test/java/org/apache/ignite/internal/sql/engine/planner/AbstractPlannerTest.java (hunk 1 above)
Comment (korlov42) [reply modules/sql-engine/src/test/java/org/apache/ignite/internal/sql/engine/planner/AbstractPlannerTest.java:573]: well, it's a bit messy... This variable is just a field of an abstract class. Every test in test class extending `AbstractPlannerTest` will overwrite this variable if validation has failed. So, this is the last error message for test class, but this is a first and only error message for test.

File: modules/sql-engine/src/test/java/org/apache/ignite/internal/sql/engine/planner/datatypes/NumericComparisonTypeCoercionTest.java (hunk 2 above)
Comment (korlov42) [reply modules/sql-engine/src/test/java/org/apache/ignite/internal/sql/engine/planner/datatypes/NumericComparisonTypeCoercionTest.java:724]: that's true, but I would prefer to leave this as is. 

To wrote all those test cases, first, I defined a list of types that I would like to verify. Among others, I definitely did want to verify different DECIMAL types: exactly the same types, types with equal precision but different scale, types with equal scale but different precision, completely different types. Then, I made a Cartesian join of the type list a leave only those pairs where first type is lesser then or equals to second. That's how I end up with *duplicates. Technically, it's possible to exclude such pairs, but it much easier and  safer just to include them as valid test cases. After all, every test case takes couple of milliseconds only. Not a big deal, I believe.

**Pull request merged**